pkg/apis/pipeline/v1beta1/taskrun_types.go

```diff
@@ -385,18 +385,35 @@ func (tr *TaskRun) IsCancelled() bool {
 
 // HasTimedOut returns true if the TaskRun runtime is beyond the allowed timeout
 func (tr *TaskRun) HasTimedOut(ctx context.Context) bool {
-	if tr.Status.StartTime.IsZero() {
+	startTime := tr.getStartTime()
+	if startTime.IsZero() {
 		return false
 	}
 	timeout := tr.GetTimeout(ctx)
 	// If timeout is set to 0 or defaulted to 0, there is no timeout.
 	if timeout == apisconfig.NoTimeoutDuration {
 		return false
 	}
-	runtime := time.Since(tr.Status.StartTime.Time)
+	runtime := time.Since(startTime.Time)
 	return runtime > timeout
 }
 
+func (tr *TaskRun) getStartTime() *metav1.Time {
```

Comment:
what's a bit confusing for me about this function is that it relies on RetriesStatus already having been updated, and now you have to rely on the `getStartTime` method instead of `Status.StartTime`. What could make more sense is updating the `tr.Status.StartTime` at the same time as `RetriesStatus` is updated, and then tr.Status.StartTime can be used directly. WDYT?

Reply:
Oh ok I understand now, thanks! Can you comment this method to make that a bit more clear?

Reply:
I agree with @lbernick, instead of iterating over the `RetriesStatus`, can we rely on `Status.StartTime` such that `Status.StartTime` is not updated with every retry and kept it as original start time?

Reply:
@pritidesai that's one of the alternatives considered, but we need to know the execution time of each attempt 

> not resetting the start time of the `TaskRun` upon retry, however it will be challenging to know the execution times of each retry

Reply:
> What could make more sense is updating the `tr.Status.StartTime` at the same time as `RetriesStatus` is updated

thought @lbernick was suggesting updating the `startTime` instead of keeping it as the original `startTime` - which is what we already do - please correct me if I misunderstood that

Reply:
maybe we can go with the second alternative where we record the actual start time as an implementation detail and we use that to compare against the timeout in `HasTimedOut` function, instead of checking the `RetriesStatus`

> keeping track of the actual start time in an extra field, but this is information that's already available in the status

Reply:
my understanding is, there are multiple `StartTime` available as part of the taskRun status:

*  `tr.Status.StartTime`
*  a list of `tr.Status.RetriesStatus[].StartTime`

Is `tr.Status.StartTime` changing with each new retry?

Reply:
> Is `tr.Status.StartTime` changing with each new retry?

@pritidesai yes, that's correct 

when we do a retry, we add the status to retry history in line 771 then we clear the status is line 772

https://github.com/tektoncd/pipeline/blob/28f950700e99fd22a175eab7e2c803248675cca0/pkg/reconciler/pipelinerun/pipelinerun.go#L763-L779

when you look into what clear status does, you see that it sets the `tr.Status.StartTime` to `nil` in line 945
https://github.com/tektoncd/pipeline/blob/28f950700e99fd22a175eab7e2c803248675cca0/pkg/reconciler/pipelinerun/pipelinerun.go#L944-L948

Reply:
I also feel like it's a bit odd that `getStartTime` will return a different value than if you accessed the start time directly - it feels like we're changing the meaning of "start time" so that it now means 2 different things

on the other hand given the way retries are implemented (reusing the same Taskrun) this does make a sort of sense, and has the advantage of containing the logic for this check in one place...

A couple bits of possible food for thought:
* Runs are going to need this same logic if since iirc [TEP-0069](https://github.com/tektoncd/community/blob/main/teps/0069-support-retries-for-custom-task-in-a-pipeline.md) is going to use the same logic for Runs - which makes me wonder if it makes sense to put this logic somewhere that can be leveraged by both vs. as a function on the TaskRun object?
* What would this logic look like if we DID have multiple TaskRuns created when retrying? We wouldn't be able to overload the start time lookup of one individual taskRun, i think we'd need to look up the previous attempts and find the earliest start time (or keep that information somewhere else as suggested in the alternatives)

The above 2 thoughts have me thinking that TaskRun might not be the best place for this logic but I don't feel strongly about it!